Assets/Mirage/ObjectPolling/PooledObject.cs

```diff
@@ -0,0 +1,262 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Mirage.Runtime
+{
+    public struct PooledObject<T>
+    {
+        public T Object;
+        public NetworkIdentity NetId;
+    }
+
+    internal struct ObjectPooling<T> where T : Object
+    {
+        #region Fields
+
+        private readonly Queue<PooledObject<T>> _inactiveGameObjects;
+        private readonly T _objectPrefabToPool;
+        private readonly NetworkIdentity _networkObjectPrefabToPool;
+        internal readonly HashSet<int> PooledObjectIDs;
+        private readonly GameObject _parentBackObject;
+
+        #endregion
+
+        /// <summary>
+        ///     Constructor for this object pool.
+        /// </summary>
+        /// <param name="poolObject">The object we want to have pooled.</param>
+        /// <param name="initialPool">How many objects do we want to start out with in the pool.</param>
+        /// <param name="networkParentObject">When sending back to pool, where should we parent the object back to.</param>
+        public ObjectPooling(T poolObject, int initialPool, GameObject networkParentObject)
+        {
+            _objectPrefabToPool = poolObject;
+            _inactiveGameObjects = new Queue<PooledObject<T>>(initialPool);
+            PooledObjectIDs = new HashSet<int>();
+            _parentBackObject = networkParentObject;
+            _networkObjectPrefabToPool = null;
+        }
+
+        /// <summary>
+        ///     Constructor for this object pool.
+        /// </summary>
+        /// <param name="poolObject">The network identity we want to have pooled.</param>
+        /// <param name="initialPool">How many objects do we want to start out with in the pool.</param>
+        /// <param name="networkParentObject">When sending back to pool, where should we parent the object back to.</param>
+        public ObjectPooling(NetworkIdentity poolObject, int initialPool, GameObject networkParentObject)
+        {
+            _networkObjectPrefabToPool = poolObject;
+            _inactiveGameObjects = new Queue<PooledObject<T>>(initialPool);
+            PooledObjectIDs = new HashSet<int>();
+            _parentBackObject = networkParentObject;
+            _objectPrefabToPool = null;
+        }
+
+        #region Network Pool Spawning
+
+        /// <summary>
+        ///     Method to spawn network object from our inactive pool.
+        /// </summary>
+        /// <param name="position">The position we want to spawn the object at.</param>
+        /// <param name="rotation">The rotation we want to have the object spawn with.</param>
+        /// <returns></returns>
+        public NetworkIdentity NetworkSpawn(Vector3 position, Quaternion rotation)
+        {
+            while (true)
```

Comment:
does this need to be a while loop. it is confusing and hard to see what the loop is actually doing

Reply:
It does because if we call to get an object from pool and we don't find one in pool at the time it loops back to try again or create a new pool object. It could be miliseconds get a new object back and this will be alot better then it failing and not spawning an object.